app/models/manageiq/providers/openstack/cloud_manager.rb

```diff
@@ -40,6 +40,21 @@ def self.default_blacklisted_event_names
     )
   end
 
+  def hostname_uniqueness_valid?
+    return unless hostname_required?
+    return unless hostname.present? # Presence is checked elsewhere
+
+    existing_providers = Endpoint.where(:hostname => hostname.downcase).where.not(:resource_id => id)
+                                 .includes(:resource).select { |endpoint| endpoint.resource.uid_ems == uid_ems }
```

Comment:
If I count right, this does only 2 queries. I think it should be good performance wise. The readability is better than the JOIN. 

Btw. there is uid_ems alias called keystone_v3_domain_id, could you use that? So we can see what we are comparing.